### ISSUE TITLE
Support for multiple renderers

### ISSUE THREAD

**Issue by aleemb** [thephpleague/commonmark#35]
My use case is that I need custom renderers for images and depending on the image URL, these renders will render the image as an iframe for YouTube, Vimeo, etc. At present I use the Sundown renderer and I have stuffed all these renderers into one big file but would like to move over to this lib and cleanup the mess while I'm at it.

I also imagine that a lot of contributors can then share their own YoutubeRenderer (or generic VideoRenderer). This would make extensibility a lot more fun and build an ecosystem around it. Maybe even give way to an extensions repo.

Based on your suggestion in #32, I currently render my custom Renderer as

```
$environment->addInlineRenderer('League\CommonMark\Inline\Element\Image', new MediaRenderer());
```

However, this only allows attaching a single renderer. It would be nice if the renderer can work sort of like conventional event handlers with the possibility to either chain the output as input to other renderers or stop propagation. The easiest approach would be to have symmetric input/output for renderers that would make chaining a breeze. Currently the input/output is `(AbstractInline, HtmlRenderer)/HtmlElement`.

**Comment by indrora** [comment]
Taking on my $0.002.

I'm attempting to implement a new kind of link type (internal references) and I'm having a hard time. I'm trying to make it such that ```^[some page](some/where/inside)``` can be parsed into what would be ```[some page](/some/path/some/where/inside)```

Having a way to mangle links like this would be *fantastic* -- it would give the ability to have what OP wants (Totally override the link process) but also fit several other use cases (e.g. a custom one to make your internal documentation, say `$[this CVE](2008-4250)` could become the equivalent of `[this CVE](http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2008-4250)`

**Comment by bfrohs** [comment]
I implemented YouTube video rendering using a processor that would replace any `Link` that has a YouTube video URL with a `YouTubeVideo` element that extends `Link`, leaving the url, label, and link alone as they're passed through. Then I added a `YouTubeVideoRenderer` that ensures the link is by itself in a paragraph, pulls the video ID out of the URL, and creates the necessary elements to display the video.

This way, if the YouTube processor/render are ever removed, it will fallback to a link to the YouTube video in it's own paragraph, which doesn't look half bad and still serves the same purpose.

See https://gist.github.com/bfrohs/9017f38586404f161298 for code.

**Comment by skyzyx** [comment]
IMO, Option 2 is "the right way" to do it. Fire an event when something happens in the parsing/rendering, and register listeners in a chain that can modify the subject as necessary. Subject/Observer Pattern.

**Comment by colinodell** [comment]
My primary objection to Option 1 was ensuring that developers could successfully influence the order in which the renderers executed. However, once #210 is merged, extensions will be loaded in the order they were added (instead of by name), thus giving the control needed for this to work.  So if you wanted to override a core renderer, you'd have to ensure your extension is loaded before the code extension.  

(This would require additional work outside of #210; that PR simply sets some of the ground work)

**Comment by colinodell** [comment]
Please review #341 when you get a chance - I'd love to hear your feedback!

**Comment by colinodell** [comment]
Implemented via #341
